pkg/controller/acmeorders/sync.go

```diff
@@ -505,24 +506,46 @@ func (c *controller) finalizeOrder(ctx context.Context, cl acmecl.Interface, o *
 		derBytes = block.Bytes
 	}
 
+	// Call to CreateOrderCert finalizes the ACME order. This call can only be made once.
 	certSlice, certURL, err := cl.CreateOrderCert(ctx, o.Status.FinalizeURL, derBytes, true)
-	// if an ACME error is returned and it's a 4xx error, mark this Order as
-	// failed and do not retry it until after applying the global backoff.
-	if acmeErr, ok := err.(*acmeapi.Error); ok {
-		if acmeErr.StatusCode >= 400 && acmeErr.StatusCode < 500 {
-			log.Error(err, "failed to finalize Order resource due to bad request, marking Order as failed")
-			c.setOrderState(&o.Status, string(cmacme.Errored))
-			o.Status.Reason = fmt.Sprintf("Failed to finalize Order: %v", err)
-			return nil
+
+	acmeErr, ok := err.(*acmeapi.Error)
+
+	// If finalizing the order returns a 4xx error then either we have
+	// already finalized this order or the order can be considered failed.
+	if ok && acmeErr.StatusCode >= 400 && acmeErr.StatusCode < 500 {
+
+		// Check if finalizing the ACME order failed because it is
+		// already valid. This scenario is possible if the ACME order
+		// has already been finalized in an earlier reconcile, but the
+		// reconciler failed to update the status of the Order CR.
+		acmeOrder, getOrderErr := getACMEOrder(ctx, cl, o)
```

Comment:
We could have also checked whether the ACME order is valid _before_ attempting to finalize. However, this would again result in an extra call to ACME server in the majority of cases where the call to `CreateOrderCert` succeeds.

Reply:
I agree.

I am thinking... #4612 does not particularly mention issues with individual calls ("fetch order", or "finalize"), but seems to suggest that calls to "new-order" are too frequent.

I wonder, are we optimizing the correct thing here?

Reply:
I reduced the [1-hour delay between retries](https://github.com/irbekrm/cert-manager/blob/ec1bdc4983a1a91ad381f356135e2f93b79aa505/pkg/controller/certificates/util.go#L41) to 10 seconds.

We can see that `new-order` is called even though the ACME server returns a permanent/non-retryable error (`rejectedIdentifier`):
![Screenshot from 2022-01-11 13-41-13](https://user-images.githubusercontent.com/2195781/148944425-883815fc-c9e0-4ce1-ab5b-669d1d3595bd.png)
 
 I imagine that people with 10,000 certificates will generate 10,000 `new-order` calls every hour even though they should not.

Reply:
Agree, this is something we should definitely work on 👍🏼 I also want to look into exponential backoff for CertificateRequests- these two probably need to be looked at together